authorization/invitation/model/invitation_model_service.go

```diff
@@ -0,0 +1,201 @@
+		if !authorization.CanHaveMembers(identityResource.ResourceType.Name) {
+			return errors.NewInternalErrorFromString(ctx, "may only invite a user to an organization, team or security group")
+		}
+	} else if inviteToResource != nil {
+		// Confirm that the issuing user has the manage members scope for the resource
+		scope, err := s.permModelSvc.HasScope(ctx, issuingUserId, inviteToResource.ResourceID, authorization.ManageMembersScope)
+		if err != nil {
+			return errors.NewInternalError(ctx, err)
+		}
+
+		if !scope {
+			return errors.NewForbiddenError(fmt.Sprintf("user requires %s scope to invite other users", authorization.ManageMembersScope))
+		}
+	}
+
+	// Iterate through all of the invitations and confirm that for each one:
+	// 1) a valid user has been specified via its User ID, e-mail address or username
+	// 2) any roles specified are valid roles for the organization, team or security group
+	// For each invitation, ensure that the IdentityID value can be found and set it
+	for i, invitation := range invitations {
+		// If the UserID has been provided, confirm it is valid and that the identity is a user
+		if invitation.IdentityID != nil {
+			identity, err := repo.Identities().Load(ctx, *invitation.IdentityID)
+			if err != nil {
+				return errors.NewInternalErrorFromString(ctx, fmt.Sprintf("invalid identity ID specified: %s", invitation.IdentityID))
+			}
+
+			if !identity.IsUser() {
+				return errors.NewBadParameterErrorFromString("Identity ID", invitation.IdentityID, "identity is not a user")
+			}
+		} else if invitation.UserName != nil {
+			// If the username has been provided, confirm the user is valid and that the identity is a user, and set the UserID
+			identities, err := repo.Identities().Query(account.IdentityFilterByUsername(*invitation.UserName))
+			if err != nil {
+				return errors.NewInternalError(ctx, err)
+			}
+			// If there is anything other than 1 result found, we have a problem
+			if len(identities) == 0 {
+				// If no users are found, return an error
+				return errors.NewBadParameterErrorFromString("Username", invitation.UserName, "username not found")
+			}
+
+			// Set the IdentityID to that of the identity found
+			invitations[i].IdentityID = &identities[0].ID
+		} else if invitation.UserEmail != nil {
+			// If the user's e-mail address has been provided, confirm the user is valid and that the identity is a user, and set the UserID
+			users, err := repo.Users().Query(account.UserFilterByEmail(*invitation.UserEmail))
+			if err != nil {
+				return errors.NewInternalError(ctx, err)
+			}
+			// We expect exactly 1 user to be found, if not we return an error
+			if len(users) == 0 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "user with e-mail address not found")
+			} else if len(users) > 1 {
+				return errors.NewBadParameterErrorFromString("E-mail", invitation.UserEmail, "more than one user with e-mail address found")
```

Comment:
It can't happen. User emails are unique in the user table.

Reply:
This section has been removed